network/hostport.go

```diff
@@ -236,3 +238,115 @@ func EnsureFirstHostPort(first HostPort, hps []HostPort) []HostPort {
 	result = append([]HostPort{first}, result...)
 	return result
 }
+
+// UniqueHostPorts sends the unique entries in the given hostPorts slice on the
+// returned channel, closing the channel when done. It takes a stop channel,
+// which can be used to signal the sending loop to stop earlier.
+func UniqueHostPorts(stop <-chan struct{}, hostPorts []HostPort) <-chan HostPort {
```

Comment:
It doesn't seem like a host would ever have so many addresses that uniquifying them would need to be async (it'll be much much faster than a network hop anyway). This would be better just returning `[]HostPort` and not using the goroutine.